### DISCUSSION TITLE
Appending mapped relationships to an existing table mapping.

### DISCUSSION THREAD

**Discussion by jenstroeger** [sqlalchemy/sqlalchemy#11124] [Usage Questions] [answered]
My question is a continuation of the section [Appending additional columns to an existing Declarative mapped class](https://docs.sqlalchemy.org/en/20/orm/declarative_tables.html#appending-additional-columns-to-an-existing-declarative-mapped-class) from the manual. That section discusses appending _columns_ but in my case I need to add a relationship mapper.

For example:
```python
class Base(DeclarativeBase):
    pass

class B(Base):
    __tablename__ = "b"

    id: Mapped[int] = mapped_column(sa.Integer(), primary_key=True)
    a_id: Mapped[int] = mapped_column(sa.Integer(), sa.ForeignKey("a.id"), nullable=False)

    all_a: Mapped["A"] = relationship("A", back_populates="all_b")
    # active_a should go here!

class A(Base):
    __tablename__ = "a"

    StatusEnum = enum.Enum("StatusEnum", ["ACTIVE", "INACTIVE", "PENDING"], start=0)

    id: Mapped[int] = mapped_column(sa.Integer(), primary_key=True)
    status: Mapped[StatusEnum] = mapped_column(sa.Enum(StatusEnum), nullable=False, default=StatusEnum.PENDING)

    all_b: Mapped[list["B"]] = relationship("B", back_populates="all_a")
```
I’d like to have an attribute `B.active_a` which maps the one `"ACTIVE"` row from table `"a"` to a `B` object. The problem is that `B` needs access to `A.StatusEnum` and I can’t rearrange the two classes (not in the real code). So I added below `A` the following:
```python
B.active_a: Mapped["A"] = relationship(
    "A", primaryjoin=f"and_(a.c.id == b.c.a_id, a.c.status == '{A.StatusEnum.ACTIVE.name}')", viewonly=True
)
```
This seems to work although `mypy` complains:
```
error: Type cannot be declared in assignment to non-self attribute  [misc]
    B.active_a: Mapped["A"] = relationship(
```
Nonetheless, I wanted to make sure, because the section in the manual states that

> A declarative table configuration allows the addition of new [Column](https://docs.sqlalchemy.org/en/20/core/metadata.html#sqlalchemy.schema.Column) objects to an existing mapping […]

but doesn’t mention such a relationship mapping.

**Comment by zzzeek** [comment, marked as answer]
it works for relationships equally well examples are here: https://docs.sqlalchemy.org/en/20/orm/basic_relationships.html#adding-relationships-to-mapped-classes-after-declaration

**Reply by jenstroeger** [reply]
Hot diggity! That’s the section I was looking for 👍🏼 Maybe you can cross-reference both sections with each other?

@zzzeek as for the type error, I silenced that. Do you have any other recommendations? But considering that the docs say

> (noting that annotated forms are not supported in this context)

it’s probably safe to silence the error.

**Reply by zzzeek** [reply]
I can accept a PR that cross references them if you know how to do that

**Reply by jenstroeger** [reply]
> I can accept a PR that cross references them if you know how to do that

See PR https://github.com/sqlalchemy/sqlalchemy/pull/11133.
